### ISSUE TITLE
Multiple Address Validation will create large countryCollection queries

### ISSUE THREAD

**Issue by Rorando** [magento/magento2#18304]
### Summary 
Calling `getCountryCollection()->getAllIds()` on `\Magento\Directory\Helper\Data` multiple times will generate duplicate filters on the collection for each call

### Preconditions
1. Magento 2.2.6 (Clean composer installation)
2. PHP 7.0.30-1 / Nginx 1.13.9 / MySql 5.6.37

### Steps to reproduce
1. Inject \Magento\Directory\Helper\Data $directoryData
2. Call `$directoryData->getCountryCollection()->getAllIds()` multiple times

### Expected result
I expect that the query from each call would be the same, or that only one query would be executed and the other calls retrieve it from stored results

### Actual result
For every call an country_id filter is added to the query with the filter from the previous call still set.
Meaning that is you call it 10 times the last query will have the same filter added 10 times.

`SELECT main_table.country_id FROM directory_country AS main_table 

WHERE (country_id IN('AF', 'AL', 'DZ', 'AS', 'AD', 'AO', 'AI', 'AQ', 'AG', 'AR', 'AM', 'AW', 'AU', 'AT')) 
  AND (country_id IN('AF', 'AL', 'DZ', 'AS', 'AD', 'AO', 'AI', 'AQ', 'AG', 'AR', 'AM', 'AW', 'AU', 'AT')) 
  AND (country_id IN('AF', 'AL', 'DZ', 'AS', 'AD', 'AO', 'AI', 'AQ', 'AG', 'AR', 'AM', 'AW', 'AU', 'AT')) 
  AND (country_id IN('AF', 'AL', 'DZ', 'AS', 'AD', 'AO', 'AI', 'AQ', 'AG', 'AR', 'AM', 'AW', 'AU', 'AT'))
......`

### Investigations so far

The getCountryCollection() does have a check if the collection is loaded `isLoaded()` only the `loadByStore()` call doesn't set the collection as loaded.

Setting the collection as loaded on the `loadByStore()` function would fix it, so would resetting the collection before the loadByStore call.
But i don't think these 2 are the best possible sollutions.

**Comment by magento-engcom-team** [comment]
Hi @Rorando. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@Rorando do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by Rorando** [comment]
@magento-engcom-team 

> Hi @Rorando. Thank you for your report.
> To help us process this issue please make sure that you provided the following information:
 * [X]  Summary of the issue
 * [X]  Information on your environment
 * [X]  Steps to reproduce
 * [X]  Expected and actual results



> @Rorando do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?

 * [X]  yes

**Comment by yaroslav-zenin** [comment]
> The getCountryCollection() does have a check if the collection is loaded isLoaded() only the loadByStore() call doesn't set the collection as loaded.

That's strange because I've checked code on 2.2.6 tag and there are present condition
```php
namespace Magento\Directory\Helper;
class Data extends \Magento\Framework\App\Helper\AbstractHelper
{
    public function getCountryCollection($store = null)
    {
        if (!$this->_countryCollection->isLoaded()) {
            $this->_countryCollection->loadByStore($store);
        }
        return $this->_countryCollection;
    }
}
```

**Comment by Rorando** [comment]
@yaroslav-zenin 
> That's strange because I've checked code on 2.2.6 tag and there are present condition

It does have an isLoaded() check, but the loadByStore function does not set the collection as loaded.
The isLoaded() check should also be store.specific then instead of how it is now

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [x] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [x] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [x] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [x] 6. Add label `Issue: Confirmed` once verification is complete. 

- [x] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by magento-engcom-team** [comment]
@engcom-backlog-nazar Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-96707`, `MAGETWO-96708` were created

**Comment by magento-engcom-team** [comment]
@Rorando, thank you for your report.     
    
Unfortunately, we are archiving this ticket now as it did not get much attention from both Magento Community and Core developers for an extended period. This is done in an effort to create a quality, community-driven backlog which will allow us to allocate the required attention more easily.

Please feel free to comment, reopen or create new ticket according to the [Issue reporting guidelines](https://github.com/magento/magento2/wiki/Issue-reporting-guidelines) 
if you are still facing this issue on the latest `2.3-develop` branch. Thank you for collaboration.
